lib/AST/DeclContext.cpp

```diff
@@ -1134,14 +1095,46 @@ DeclContextKind DeclContext::getContextKind() const {
   llvm_unreachable("Unhandled DeclContext ASTHierarchy");
 }
 
+#define DECL(Id, Parent) \
+  static_assert(!std::is_base_of<DeclContext, Id##Decl>::value, \
+                "Non-context Decl node has context?");
+#define CONTEXT_DECL(Id, Parent) \
+  static_assert(alignof(DeclContext) == alignof(Id##Decl), "Alignment error"); \
+  static_assert(std::is_base_of<DeclContext, Id##Decl>::value, \
+                "CONTEXT_DECL nodes must inherit from DeclContext");
+#define CONTEXT_VALUE_DECL(Id, Parent) \
+  static_assert(alignof(DeclContext) == alignof(Id##Decl), "Alignment error"); \
+  static_assert(std::is_base_of<DeclContext, Id##Decl>::value, \
+                "CONTEXT_VALUE_DECL nodes must inherit from DeclContext");
+#include "swift/AST/DeclNodes.def"
+
+#define EXPR(Id, Parent) \
+  static_assert(!std::is_base_of<DeclContext, Id##Expr>::value, \
+                "Non-context Expr node has context?");
+#define CONTEXT_EXPR(Id, Parent) \
+  static_assert(alignof(DeclContext) == alignof(Id##Expr), "Alignment error"); \
+  static_assert(std::is_base_of<DeclContext, Id##Expr>::value, \
+                "CONTEXT_EXPR nodes must inherit from DeclContext");
+#include "swift/AST/ExprNodes.def"
+
+#ifndef NDEBUG
+// XXX -- static_cast is not static enough for use with static_assert().
+// DO verify this by temporarily breaking a Decl or Expr.
+// DO NOT assume that the compiler will emit this code blindly.
+__attribute__((constructor))
+static void verify_DeclContext_is_start_of_node() {
+  auto decl = reinterpret_cast<Decl*>(sizeof(DeclContext));
 #define DECL(Id, Parent)
 #define CONTEXT_DECL(Id, Parent) \
-  static_assert(alignof(DeclContext) == alignof(Id##Decl), "Alignment error");
+  assert(nullptr == static_cast<DeclContext*>(static_cast<Id##Decl*>(decl)));
```

Comment:
I don't trust the compiler not to optimize this away, even at -O0. I can't decide if I'd rather have this or no checking at all, but it would be slightly better if you `reinterpret_cast`ed back to uintptr_t and compared against 0 rather than using `nullptr`.

Reply:
I've verified that this `assert` is not optimized away at -O3. That is why I put the comment above about *actually* testing the code.

Observationally speaking, the `assert` check is non-constant at the front end, but by the time the backend is done with it, either the assert will fire unconditionally, or it compiles away.

Reply:
I don't really trust a future compiler, or a non-Clang compiler, to always pick "never assert" or "dynamically check". Avoiding testing against `nullptr` would make me feel slightly better.

(Also, on second thought, this probably goes better in, say, the AST verifier, rather than as a static initializer.)

Reply:
I can change the check if you want. Personally, I think this code is more future/compiler proof than you might think. The real undefined-behavior-turns-into-deleted-code risk isn't the `nullptr` comparison, but that the cast to `GenericContext` can go wrap around the address space temporarily before being cast back to a `DeclContext`. I can fix that with a different fake starting pointer (like 0x1000 + sizeof(DeclContext) and checking that the net result of the casting equals 0x1000.

As for the AST verifier, that won't work in practice because it runs way too late. Stuff will have blown up long before then if a DeclContext isn't properly inherited from.

Reply:
The address space can't "wrap around" because pointers don't have an integer representation in the C standard. All pointer arithmetic must be in-bounds, and NULL is never a valid pointer. That said, picking a different starting location is probably a good idea anyway.

Reply:
(thanks for humoring my paranoia)

Reply:
No problem-o.